server.js

```diff
@@ -15,40 +24,88 @@ app.use("/public", express.static(__dirname + "/public"));
 app.use(cookieParser());
 
 // get requests
-app.get("/", (req, res) => {
+app.get("/", async (req, res) => {
     const { cookies } = req
-    res.render('index', sessions[cookies.session_id]);
+    try {
+        const currentSession = await Session.findById(cookies.session_id)
+        if (currentSession === null) {
+            res.render('index');
+        } else {
+            res.render('index', currentSession);
+        }
+
+    } catch (err) {
+        res.status(500).json({ result: 'bad', error: err.message })
+    }
+
 })
 // post requests
 app.post("/signup", async (req, res) => {
+    const signupUser = await User.findOne({ username: req.body.username })
+    try {
+        if (signupUser) {
+            console.log("user not added");
+            res.redirect('/');
+        } else {
+            const newUser = await User.create({
+                username: req.body.username,
+                password: req.body.password,
+            });
+            console.log("user added");
+            res.redirect('/');
+        }
 
-    users[req.body.username] = req.body.password;
-    res.redirect('/');
+    } catch (err) {
+        res.status(500).json({ result: 'bad', error: err.message })
```

Comment:
The same route should not sometimes return HTML and sometimes return JSON. Either you always return JSON (and use client-side JavaScript do do whatever you want with it), or you render an HTML page (in the case of an error, render an error template)